lib/request/RequestQueueV2.js

```diff
@@ -63,36 +66,39 @@ RequestQueueV2.prototype = {
             // The request has been sent, attempt to jump on the request
             // if possible.
             if (request.sent) {
-                var results = request.add(rRemainingPaths, oRemainingPaths, refCountCallback);
-
-                // Checks to see if the results were successfully inserted
-                // into the outgoing results.  Then our paths will be reduced
-                // to the complement.
-                if (results[0]) {
-                    rRemainingPaths = results[1];
-                    oRemainingPaths = results[2];
-                    disposables[disposables.length] = results[3];
-                    ++count;
+                if (this.model._enableRequestDeduplication) {
+                    var results = request.add(rRemainingPaths, oRemainingPaths, refCountCallback);
+
+                    // Checks to see if the results were successfully inserted
+                    // into the outgoing results.  Then our paths will be reduced
+                    // to the complement.
+                    if (results[0]) {
+                        rRemainingPaths = results[1];
+                        oRemainingPaths = results[2];
+                        disposables[disposables.length] = results[3];
+                        ++count;
+
+                        // If there are no more remaining paths then exit the loop.
+                        if (!oRemainingPaths.length) {
```

Comment:
could it be useful to change request.add to potentially return `null` instead of empty array for `rRemainingPaths` and `oRemainingPaths`, just like you've done for them below.

Reply:
I'm unsure what you mean; `request.add` is not called when request deduplication is disabled. I made the switch to `null` assignments on lines 92-93 to avoid the allocation of unused empty arrays.

Reply:
aligned offline that gains from optimizing array allocations inside of `request.add` will be a larger undertaking.